File: app/schemas/reports.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class OperationalReservationSummary(BaseModel):
    reservation_id: int
    confirmation_code: str
    guest_id: int
    guest_name: str | None = None
    room_id: int | None = None
    room_number: str | None = None
    status: str
    check_in_date: date
    check_out_date: date
    total_amount: Decimal | None = None
    amount_paid: Decimal | None = None
    balance_due: Decimal | None = None


class OperationalReservationGroup(BaseModel):
    count: int
    reservations: list[OperationalReservationSummary] = Field(default_factory=list)


class AvailableWithReviewItem(BaseModel):
    reservation_id: int
    confirmation_code: str
    room_id: int | None = None
    room_number: str | None = None
    check_in_date: date
    check_out_date: date
    allocation_status: str


class ActiveRoomBlockItem(BaseModel):
    room_block_id: int
    room_id: int
    room_number: str | None = None
    reason_code: str
    reason_note: str | None = None
    starts_at: date
    ends_at: date | None = None
    is_indefinite: bool


class CashSessionStatusRead(BaseModel):
    status: str
    session_id: int | None = None
    opened_at: datetime | None = None
    opened_by_user_id: int | None = None
    currency_code: str | None = None


class OperationalAlertRead(BaseModel):
    code: str
    severity: str
    message: str
    reservation_id: int | None = None
    room_id: int | None = None
    room_block_id: int | None = None
    cash_session_id: int | None = None


class DailyOperationalReportRead(BaseModel):
    hotel_id: int
    report_date: date
    generated_at: datetime
    arrivals: OperationalReservationGroup
    departures: OperationalReservationGroup
    pending_payments: OperationalReservationGroup
    late_arrivals: list[OperationalReservationSummary] = Field(default_factory=list)
    available_with_review: list[AvailableWithReviewItem] = Field(default_factory=list)
    active_room_blocks: list[ActiveRoomBlockItem] = Field(default_factory=list)
    cash_session: CashSessionStatusRead
    alerts: list[OperationalAlertRead] = Field(default_factory=list)


class NightlyOperationalSummaryRead(BaseModel):
    hotel_id: int
    report_date: date
    generated_at: datetime
    alert_count: int
    pending_payment_count: int
    late_arrival_count: int
    available_with_review_count: int
    active_room_block_count: int
    cash_session_status: str
    alerts: list[OperationalAlertRead] = Field(default_factory=list)


class OperationalReportDeliveryRead(BaseModel):
    delivered: bool
    channel: str | None = None
    sender_email: str | None = None
    provider_message_id: str | None = None
    recipients: list[str] = Field(default_factory=list)

File: frontend/src/api/reports.ts
import { apiFetch, type SessionLike } from "./client";

export type OperationalReservationSummary = {
  reservation_id: number;
  confirmation_code: string;
  guest_id: number;
  guest_name?: string | null;
  room_id?: number | null;
  room_number?: string | null;
  status: string;
  check_in_date: string;
  check_out_date: string;
  total_amount?: number | null;
  amount_paid?: number | null;
  balance_due?: number | null;
};

export type OperationalReservationGroup = {
  count: number;
  reservations: OperationalReservationSummary[];
};

export type AvailableWithReviewItem = {
  reservation_id: number;
  confirmation_code: string;
  room_id?: number | null;
  room_number?: string | null;
  check_in_date: string;
  check_out_date: string;
  allocation_status: string;
};

export type ActiveRoomBlockItem = {
  room_block_id: number;
  room_id: number;
  room_number?: string | null;
  reason_code: string;
  reason_note?: string | null;
  starts_at: string;
  ends_at?: string | null;
  is_indefinite: boolean;
};

export type CashSessionStatusRead = {
  status: string;
  session_id?: number | null;
  opened_at?: string | null;
  opened_by_user_id?: number | null;
  currency_code?: string | null;
};

export type OperationalAlert = {
  code: string;
  severity: string;
  message: string;
  reservation_id?: number | null;
  room_id?: number | null;
  room_block_id?: number | null;
  cash_session_id?: number | null;
};

export type DailyOperationalReport = {
  hotel_id: number;
  report_date: string;
  generated_at: string;
  arrivals: OperationalReservationGroup;
  departures: OperationalReservationGroup;
  pending_payments: OperationalReservationGroup;
  late_arrivals: OperationalReservationSummary[];
  available_with_review: AvailableWithReviewItem[];
  active_room_blocks: ActiveRoomBlockItem[];
  cash_session: CashSessionStatusRead;
  alerts: OperationalAlert[];
};

export type NightlyOperationalSummary = {
  hotel_id: number;
  report_date: string;
  generated_at: string;
  alert_count: number;
  pending_payment_count: number;
  late_arrival_count: number;
  available_with_review_count: number;
  active_room_block_count: number;
  cash_session_status: string;
  alerts: OperationalAlert[];
};

export type OccupancyReport = {
  start_date: string;
  end_date: string;
  total_rooms: number;
  average_occupancy: number;
  daily: Array<{
    date: string;
    occupied: number;
    available: number;
    rate: number;
  }>;
};

export type RevenueReport = {
  start_date: string;
  end_date: string;
  collected: {
    total: number;
    by_method: Record<string, number>;
    by_day: Record<string, number>;
    transactions_count: number;
  };
  expected: {
    total: number;
    pending: number;
    reservations_count: number;
  };
};

export const getDailyOperationalReport = (reportDate: string, session?: SessionLike) =>
  apiFetch<DailyOperationalReport>(
    `/api/reports/operational/daily?report_date=${encodeURIComponent(reportDate)}`,
    { session }
  );

export const getOperationalAlerts = (reportDate: string, session?: SessionLike) =>
  apiFetch<NightlyOperationalSummary>(
    `/api/reports/operational/alerts?report_date=${encodeURIComponent(reportDate)}`,
    { session }
  );

export const getOccupancyReport = (startDate: string, endDate: string, session?: SessionLike) =>
  apiFetch<OccupancyReport>(
    `/api/reports/occupancy?start_date=${encodeURIComponent(startDate)}&end_date=${encodeURIComponent(endDate)}`,
    { session }
  );

export const getRevenueReport = (startDate: string, endDate: string, session?: SessionLike) =>
  apiFetch<RevenueReport>(
    `/api/reports/revenue?start_date=${encodeURIComponent(startDate)}&end_date=${encodeURIComponent(endDate)}`,
    { session }
  );
